File: students/Necross12/homework-4/src/banking_system.py
import math

AHORRO = "Ahorro"
ESTANDAR = "Estándar"
PREMIUM = "Premium"

ACTIVO = "Activo"
SUSPENDIDO = "Suspendido"
CONGELADO = "Congelado"
CERRADO = "Cerrado"


class BankAccount:
    TIPOS_VALIDOS = [AHORRO, ESTANDAR, PREMIUM]
    LIMITES = {AHORRO: 2000, ESTANDAR: 5000, PREMIUM: 50000}
    SALDO_MINIMO = {AHORRO: 100, ESTANDAR: 0, PREMIUM: 10000}

    MENSAJES_TIPO = {
        AHORRO: "Se acepta: cuenta de ahorro",
        ESTANDAR: "Se acepta: cuenta corriente",
        PREMIUM: "Se acepta: cuenta premium",
    }

    # Transiciones permitidas (Cerrado es estado final / sumidero)
    TRANSICIONES = {
        ACTIVO: [SUSPENDIDO, CONGELADO, CERRADO],
        SUSPENDIDO: [ACTIVO, CERRADO],
        CONGELADO: [ACTIVO, CERRADO],
        CERRADO: [],
    }

    MENSAJES_ESTADO = {
        ACTIVO: "Eliminar restricciones",
        SUSPENDIDO: "Enviar notificación de advertencia",
        CONGELADO: "Bloquear todas las transacciones",
        CERRADO: "Declaración final generada",
    }

    def __init__(self, account_type, initial_balance):
        self.account_type = account_type
        self.balance = initial_balance
        self.state = ACTIVO
        self.daily_transfer_total = 0
        self.last_message = ""

    @staticmethod
    def validar_tipo_cuenta(tipo):
        if tipo is None or tipo == "":
            return {"valid": False, "error": "Error: el tipo de cuenta es obligatorio"}

        if tipo not in BankAccount.TIPOS_VALIDOS:
            return {"valid": False, "error": "Error: tipo de cuenta no válido"}

        return {"valid": True, "message": BankAccount.MENSAJES_TIPO[tipo]}

    def get_daily_limit(self):
        return BankAccount.LIMITES.get(self.account_type, 0)

    def _fail(self, error):
        self.last_message = error
        return {"success": False, "error": error}

    def transfer(self, amount):
        if self.state == CONGELADO:
            return self._fail("Error: Cuenta congelada")
        if self.state == CERRADO:
            return self._fail("Error: Cuenta cerrada")
        if (
            isinstance(amount, bool)
            or not isinstance(amount, (int, float))
            or not math.isfinite(amount)
            or amount <= 0
        ):
            return self._fail("Error: el monto debe ser positivo")
        if amount > self.balance:
            return self._fail("Error: fondos insuficientes")
        if self.daily_transfer_total + amount > self.get_daily_limit():
            return self._fail("Error: supera el límite diario")

        # Redondeo a centavos para evitar errores de punto flotante
        self.balance = round(self.balance - amount, 2)
        self.daily_transfer_total = round(self.daily_transfer_total + amount, 2)

        if self.balance < BankAccount.SALDO_MINIMO[self.account_type]:
            self.state = SUSPENDIDO
        self.last_message = "La transferencia se realiza"
        return {"success": True}

    def next_state(self, target_state):
        permitidas = BankAccount.TRANSICIONES.get(self.state, [])
        if target_state not in permitidas:
            if self.state == CERRADO:
                self.last_message = "Error: Cuenta cerrada"
            else:
                self.last_message = (
                    f"Error: transición inválida de {self.state} a {target_state}"
                )
            return False
        self.state = target_state
        self.last_message = BankAccount.MENSAJES_ESTADO[target_state]
        return True

File: students/Necross12/homework-4/src/bankingSystem.js
const TIPOS = {
    AHORRO: "Ahorro",
    ESTANDAR: "Estándar",
    PREMIUM: "Premium"
};

const ESTADOS = {
    ACTIVO: "Activo",
    SUSPENDIDO: "Suspendido",
    CONGELADO: "Congelado",
    CERRADO: "Cerrado",
};

class BankAccount {
    static TIPOS_VALIDOS = [TIPOS.AHORRO, TIPOS.ESTANDAR, TIPOS.PREMIUM];

    static LIMITES = {
        [TIPOS.AHORRO]: 2000,
        [TIPOS.ESTANDAR]: 5000,
        [TIPOS.PREMIUM]: 50000,
    };

    static SALDO_MINIMO = {
        [TIPOS.AHORRO]: 100,
        [TIPOS.ESTANDAR]: 0,
        [TIPOS.PREMIUM]: 10000,
    };

    static MENSAJES_TIPO = {
        [TIPOS.AHORRO]: "Se acepta: cuenta de ahorro",
        [TIPOS.ESTANDAR]: "Se acepta: cuenta corriente",
        [TIPOS.PREMIUM]: "Se acepta: cuenta premium",
    };

    // Transiciones permitidas (Cerrado es estado final o de poso)
    static TRANSICIONES = {
        [ESTADOS.ACTIVO]: [ESTADOS.SUSPENDIDO, ESTADOS.CONGELADO, ESTADOS.CERRADO],
        [ESTADOS.SUSPENDIDO]: [ESTADOS.ACTIVO, ESTADOS.CERRADO],
        [ESTADOS.CONGELADO]: [ESTADOS.ACTIVO, ESTADOS.CERRADO],
        [ESTADOS.CERRADO]: [],
    };

    static MENSAJES_ESTADO = {
        [ESTADOS.ACTIVO]: "Eliminar restricciones",
        [ESTADOS.SUSPENDIDO]: "Enviar notificación de advertencia",
        [ESTADOS.CONGELADO]: "Bloquear todas las transacciones",
        [ESTADOS.CERRADO]: "Declaración final generada",
    };

    constructor(accountType, initialBalance) {
        this.accountType = accountType;
        this.balance = initialBalance;
        this.state = ESTADOS.ACTIVO;
        this.dailyTransferTotal = 0;
        this.lastMessage = "";
    }

    static validarTipoCuenta(tipo) {
        if (tipo === undefined || tipo === null || tipo === "")
            return { valid: false, error: "Error: el tipo de cuenta es obligatorio" };

        if (!BankAccount.TIPOS_VALIDOS.includes(tipo))
            return { valid: false, error: "Error: tipo de cuenta no válido" };

        return { valid: true, message: BankAccount.MENSAJES_TIPO[tipo] };
    }

    getDailyLimit() {
        return BankAccount.LIMITES[this.accountType] || 0;
    }

    _fail(error) {
        this.lastMessage = error;
        return { success: false, error };
    }

    transfer(amount) {
        if (this.state === ESTADOS.CONGELADO)
            return this._fail("Error: Cuenta congelada");

        if (this.state === ESTADOS.CERRADO)
            return this._fail("Error: Cuenta cerrada");

        if (typeof amount !== "number" || Number.isNaN(amount) || amount <= 0)
            return this._fail("Error: el monto debe ser positivo");

        if (amount > this.balance)
            return this._fail("Error: fondos insuficientes");

        if (this.dailyTransferTotal + amount > this.getDailyLimit())
            return this._fail("Error: supera el límite diario");

        // Redondeo
      this.balance = Number((this.balance - amount).toFixed(2));
      this.dailyTransferTotal = Number((this.dailyTransferTotal + amount).toFixed(2),);

      if (this.balance < BankAccount.SALDO_MINIMO[this.accountType])
        this.state = ESTADOS.SUSPENDIDO;

      this.lastMessage = "La transferencia se realiza";
      return { success: true };
    }

    next(targetState) {
        const permitidas = BankAccount.TRANSICIONES[this.state] || [];
        if (!permitidas.includes(targetState)) {
            this.lastMessage =
                this.state === ESTADOS.CERRADO
                    ? "Error: Cuenta cerrada"
                    : `Error: transición inválida de ${this.state} a ${targetState}`;
            return false;
        }
        this.state = targetState;
        this.lastMessage = BankAccount.MENSAJES_ESTADO[targetState];
        return true;
    }
}

module.exports = BankAccount;
module.exports.TIPOS = TIPOS;
module.exports.ESTADOS = ESTADOS;
